File: android/app/src/main/java/com/example/yugiohscanner/ml/DeckCoverage.kt
package com.example.yugiohscanner.ml

import com.example.yugiohscanner.cloud.CardRow
import com.example.yugiohscanner.cloud.ContainerRow
import com.example.yugiohscanner.cloud.CopyRow
import com.example.yugiohscanner.cloud.Deck
import com.example.yugiohscanner.cloud.DeckCard
import com.example.yugiohscanner.cloud.printingKey
import java.util.Locale

data class Reserved(val deckId: Long, val name: String, val count: Int)

data class CoverageCard(
    val cardId: String,
    val needed: Int,
    val inBox: Int,
    val available: Int,
    val missing: Int,
    val price: Double?,
    val missingCost: Double?,
    val reservedElsewhere: List<Reserved>,
)

data class CoverageTotals(
    val needed: Int, val owned: Int, val boxed: Int, val missing: Int,
    val missingCards: Int, val cost: Double, val unpriced: Int,
)

data class Coverage(val boxId: String?, val cards: List<CoverageCard>, val totals: CoverageTotals)

/**
 * Spec E1 §4/§5/§8 -- Abgleich eines Decks mit der Sammlung und die Texte dazu.
 * ZWILLING: desktop/src/utils/deckCoverage.js. Beide laufen gegen docs/fixtures/decks/coverage.json.
 * Wer eine Seite aendert, aendert beide.
 *
 * Ein Exemplar zaehlt nur, wenn es lebt UND sein Printing in [cards] steht (der Speicher fuehrt nur lebende
 * Printings; die JS-Fassung liest dafuer `printing_deleted`).
 */
object DeckCoverage {
    const val LOADING = "…"

    private fun round2(v: Double) = Math.round(v * 100.0) / 100.0

    /** Gueltige Deckbox = lebender Behaelter mit kind "deckbox". Geloescht oder umgestellt gilt als "keine Box". */
    fun validDeckboxIds(containers: List<ContainerRow>): Set<String> =
        containers.filter { !it.deleted && it.kind == "deckbox" }.mapTo(HashSet()) { it.containerId }

    /** Spec E1 §3: gueltige Deckboxen, die keinem ANDEREN Deck gehoeren, in der Reihenfolge der Behaelterliste. */
    fun deckBoxChoices(deckId: Long, decks: List<Deck>, containers: List<ContainerRow>): List<ContainerRow> {
        val taken = decks.filter { it.id != deckId }.mapNotNullTo(HashSet()) { it.containerId }
        return containers.filter { !it.deleted && it.kind == "deckbox" && it.containerId !in taken }
    }

    fun deckBoxId(deck: Deck, containers: List<ContainerRow>): String? =
        deck.containerId?.takeIf { it in validDeckboxIds(containers) }

    private class Acc(val cardId: String) {
        var needed = 0
        var inBox = 0
        var live = 0
        val reserved = LinkedHashMap<Long, Reserved>()
    }

    fun compute(
        deckId: Long,
        deckCards: List<DeckCard>,
        copies: List<CopyRow>,
        cards: List<CardRow>,
        decks: List<Deck>,
        containers: List<ContainerRow>,
        prices: Map<String, Double?>,
    ): Coverage {
        val valid = validDeckboxIds(containers)
        fun boxOf(d: Deck): String? = d.containerId?.takeIf { it in valid }
        val boxId = decks.firstOrNull { it.id == deckId }?.let { boxOf(it) }
        val otherBox = HashMap<String, Deck>()
        for (d in decks) {
            val b = boxOf(d)
            if (d.id != deckId && b != null) otherBox[b] = d
        }
        val liveKeys = cards.filter { !it.deleted }.mapTo(HashSet()) { it.printingKey() }

        val byCard = LinkedHashMap<String, Acc>()
        for (dc in deckCards) byCard.getOrPut(dc.cardId) { Acc(dc.cardId) }.needed += dc.count
        for (cp in copies) {
            if (cp.deleted || cp.printingKey() !in liveKeys) continue
            val e = byCard[cp.cardId] ?: continue
            e.live++
            val cid = cp.containerId
            if (boxId != null && cid == boxId) {
                e.inBox++
            } else if (cid != null) {
                val d = otherBox[cid] ?: continue
                val r = e.reserved[d.id] ?: Reserved(d.id, d.name, 0)
                e.reserved[d.id] = r.copy(count = r.count + 1)
            }
        }

        val deckOrder = decks.withIndex().associate { it.value.id to it.index }
        val out = ArrayList<CoverageCard>()
        var needed = 0; var owned = 0; var boxed = 0; var missingSum = 0
        var missingCards = 0; var cost = 0.0; var unpriced = 0
        for (e in byCard.values) {
            // Reihenfolge der Deckliste, damit beide Geraete die Markierungen gleich ordnen.
            val reserved = e.reserved.values.sortedBy { deckOrder[it.deckId] ?: Int.MAX_VALUE }
            val available = e.live - reserved.sumOf { it.count }
            val missing = maxOf(0, e.needed - available)
            val price = prices[e.cardId]?.takeIf { it > 0.0 }
            val missingCost = price?.let { round2(missing * it) }
            out.add(CoverageCard(e.cardId, e.needed, e.inBox, available, missing, price, missingCost, reserved))
            needed += e.needed
            owned += minOf(e.needed, available)
            boxed += minOf(e.needed, e.inBox)
            missingSum += missing
            if (missing > 0) {
                missingCards++
                if (missingCost == null) unpriced++ else cost += missingCost
            }
        }
        return Coverage(boxId, out, CoverageTotals(needed, owned, boxed, missingSum, missingCards, round2(cost), unpriced))
    }

    fun eur(v: Double): String = String.format(Locale.GERMANY, "%,.2f €", v)

    fun costText(t: CoverageTotals): String? = when {
        t.missing == 0 -> null
        t.unpriced == t.missingCards -> "Preis unbekannt"
        t.unpriced > 0 -> "ca. ${eur(t.cost)} + ${t.unpriced} ohne Preis"
        else -> "ca. ${eur(t.cost)}"
    }

    fun headerText(c: Coverage): String {
        val t = c.totals
        val cost = costText(t)?.let { " · $it" } ?: ""
        return "vorhanden ${t.owned}/${t.needed} · in der Box ${t.boxed}/${t.needed} · fehlen ${t.missing}$cost"
    }

    fun listText(c: Coverage): String {
        val t = c.totals
        val cost = costText(t)?.let { " · $it" } ?: ""
        return "${t.owned}/${t.needed} vorhanden$cost"
    }

    fun boxLabel(deck: Deck, containers: List<ContainerRow>): String {
        val id = deckBoxId(deck, containers) ?: return "keine Box"
        return containers.firstOrNull { it.containerId == id }?.name ?: "keine Box"
    }

    fun rowText(card: CoverageCard): String = "Box ${card.inBox} · verfügbar ${card.available} · gebraucht ${card.needed}"

    fun reservedTexts(card: CoverageCard): List<String> = card.reservedElsewhere.map { "${it.count} in Deck ${it.name}" }
}

File: desktop/src/utils/deckCoverage.js
import { fmtNum } from './format.js';

// Spec E1 §4/§5/§8 — Abgleich eines Decks mit der Sammlung und die Texte dazu.
// ZWILLING: android/app/src/main/java/com/example/yugiohscanner/ml/DeckCoverage.kt. Beide laufen gegen
// docs/fixtures/decks/coverage.json. Wer eine Seite aendert, aendert beide.

export const LOADING = '…';

const round2 = (v) => Math.round(v * 100) / 100;

// Gueltige Deckbox = lebender Behaelter mit kind 'deckbox'. Eine geloeschte oder umgestellte Box gilt als "keine Box".
export function validDeckboxIds(containers) {
  const out = new Set();
  for (const c of containers || []) if (c && !c.deleted && c.kind === 'deckbox') out.add(c.container_id);
  return out;
}

// Spec E1 §3: waehlbar sind gueltige Deckboxen, die keinem ANDEREN Deck gehoeren (die eigene also mit),
// in der Reihenfolge der Behaelterliste.
export function deckBoxChoices(deckId, decks, containers) {
  const taken = new Set((decks || []).filter((d) => d.id !== deckId && d.container_id).map((d) => d.container_id));
  return (containers || []).filter((c) => !c.deleted && c.kind === 'deckbox' && !taken.has(c.container_id));
}

export function deckBoxId(deck, containers) {
  const id = deck && deck.container_id;
  return id && validDeckboxIds(containers).has(id) ? id : null;
}

// input: { deckId, deckCards: [{card_id, count, section}], copies: [{copy_id, card_id, container_id, deleted?, printing_deleted?}],
//          decks: [{id, name, container_id}], containers: [{container_id, name, kind, deleted?}], prices: {passcode: number|null} }
export function deckCoverage({ deckId, deckCards, copies, decks, containers, prices }) {
  const valid = validDeckboxIds(containers);
  const boxOf = (d) => (d.container_id && valid.has(d.container_id) ? d.container_id : null);
  const own = (decks || []).find((d) => d.id === deckId);
  const boxId = own ? boxOf(own) : null;
  const otherBox = new Map(); // container_id -> Deck
  for (const d of decks || []) {
    const b = boxOf(d);
    if (d.id !== deckId && b) otherBox.set(b, d);
  }

  const byCard = new Map();
  for (const dc of deckCards || []) {
    const id = String(dc.card_id);
    const e = byCard.get(id) || { card_id: id, needed: 0, inBox: 0, live: 0, reserved: new Map() };
    e.needed += Number(dc.count) || 0;
    byCard.set(id, e);
  }
  for (const cp of copies || []) {
    if (cp.deleted || cp.printing_deleted) continue;
    const e = byCard.get(String(cp.card_id));
    if (!e) continue;
    e.live += 1;
    if (boxId && cp.container_id === boxId) e.inBox += 1;
    else if (cp.container_id && otherBox.has(cp.container_id)) {
      const d = otherBox.get(cp.container_id);
      const r = e.reserved.get(d.id) || { deck_id: d.id, name: d.name, count: 0 };
      r.count += 1;
      e.reserved.set(d.id, r);
    }
  }

  const deckOrder = new Map((decks || []).map((d, i) => [d.id, i]));
  const cards = [];
  const totals = { needed: 0, owned: 0, boxed: 0, missing: 0, missingCards: 0, cost: 0, unpriced: 0 };
  for (const e of byCard.values()) {
    // Reihenfolge der Deckliste, damit beide Geraete die Markierungen gleich ordnen.
    const reservedElsewhere = Array.from(e.reserved.values()).sort((a, b) => deckOrder.get(a.deck_id) - deckOrder.get(b.deck_id));
    const reservedCount = reservedElsewhere.reduce((s, r) => s + r.count, 0);
    const available = e.live - reservedCount;
    const missing = Math.max(0, e.needed - available);
    const p = prices ? prices[e.card_id] : null;
    const price = typeof p === 'number' && p > 0 ? p : null;
    const missingCost = price == null ? null : round2(missing * price);
    cards.push({ card_id: e.card_id, needed: e.needed, inBox: e.inBox, available, missing, price, missingCost, reservedElsewhere });
    totals.needed += e.needed;
    totals.owned += Math.min(e.needed, available);
    totals.boxed += Math.min(e.needed, e.inBox);
    totals.missing += missing;
    if (missing > 0) {
      totals.missingCards += 1;
      if (price == null) totals.unpriced += 1;
      else totals.cost += missingCost;
    }
  }
  totals.cost = round2(totals.cost);
  return { boxId, cards, totals };
}

export const eur = (v) => `${fmtNum(v)} €`;

export function costText(totals) {
  if (!totals || totals.missing === 0) return null;
  if (totals.unpriced === totals.missingCards) return 'Preis unbekannt';
  if (totals.unpriced > 0) return `ca. ${eur(totals.cost)} + ${totals.unpriced} ohne Preis`;
  return `ca. ${eur(totals.cost)}`;
}

export function headerText(cov) {
  const t = cov.totals;
  const cost = costText(t);
  return `vorhanden ${t.owned}/${t.needed} · in der Box ${t.boxed}/${t.needed} · fehlen ${t.missing}${cost ? ` · ${cost}` : ''}`;
}

export function listText(cov) {
  const t = cov.totals;
  const cost = costText(t);
  return `${t.owned}/${t.needed} vorhanden${cost ? ` · ${cost}` : ''}`;
}

export function boxLabel(deck, containers) {
  const id = deckBoxId(deck, containers);
  const c = id && (containers || []).find((x) => x.container_id === id);
  return c ? c.name : 'keine Box';
}

export const rowText = (card) => `Box ${card.inBox} · verfügbar ${card.available} · gebraucht ${card.needed}`;

export const reservedTexts = (card) => card.reservedElsewhere.map((r) => `${r.count} in Deck ${r.name}`);
